### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by mdagois**

### CODE REVIEW

**Comment by mdagois** [comment]
* Protobuf
  * Added protobuf as submodule
  * Disabled some warnings in Visual Studio
  * Enabled lib and protoc build
* Metrics
  * Added metrics manager to Application
  * Added an opt-in option to application settings
  * Added rules to convert proto files
  * Added the export function (leverages protobuf)

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -628,3 +632,34 @@ if (NOT ANDROID)
     endif()
 endif()
 
+# ------------------------------------------------------------------------------
+# Protobuf library + proto generation
+# ------------------------------------------------------------------------------
+if (NOT ANDROID)
```
Comment (Keenuts) [inline src/CMakeLists.txt:638]: Is it because for now you prefer to prototype without having to deal with it? Or is there an issue with protobuf<>android?

File: src/ppx/application.cpp [hunk 2]
```diff
@@ -668,6 +672,27 @@ void Application::DispatchShutdown()
 {
     Shutdown();
 
+    if (mSettings.useMetrics) {
+        using namespace std;
```
Comment (Keenuts) [inline src/ppx/application.cpp:676]: note: that's forbidden by the google c++ style.

(we have been using `using namespace ppx` all across the project, I know..)

File: src/proto/metrics_report.proto [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    string unit = 2;
+    Interpretation interpretation = 3;
+    Range expected_range = 4;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+message MetricCounter {
+    MetricMetadata metadata = 1;
+    int64 value = 2;
```
Comment (Keenuts) [inline src/proto/metrics_report.proto:29]: uint64?

File: src/CMakeLists.txt (hunk 1 above)
Comment (mdagois) [reply src/CMakeLists.txt:638]: Yes, there is an issue.
I use protoc to convert my proto to ***.pb.h and ***.pb.cc and it works fine when you build on the machine that is the target.
When cross-compiling for Android, protoc is built for android too... which is not really desirable.
I don't know how to fix that. :)

File: src/ppx/application.cpp (hunk 2 above)
Comment (mdagois) [reply src/ppx/application.cpp:676]: Oh, I'll change it. :)

File: src/proto/metrics_report.proto (hunk 3 above)
Comment (mdagois) [reply src/proto/metrics_report.proto:29]: Nice catch. I will fix it.

File: src/CMakeLists.txt (hunk 1 above)
Comment (Keenuts) [reply src/CMakeLists.txt:638]: the generated pb.h/cc files are not portable?

**Comment by mdagois** [comment]
> My biggest issue with this would be the proto library dependency, which bring abseil...
> But I suppose if you want to interface with perfetto that's mandatory?
> json libraries are soo much smaller, even header-only :(

Well, to be honest, when I see the pain I have to go through to make it work, I kind of agree with you haha.
And the build time is horrendous...
It was a good, and hard way, to learn, but if two people think it is a bad idea... maybe it is. :)

I will spin it back to JSON (or YAML). I can keep most of the code, since the export mechanics is just a part of the PR.

File: src/CMakeLists.txt (hunk 1 above)
Comment (mdagois) [reply src/CMakeLists.txt:638]: it seems they are. I can rebuild in gradle with generated pb.h/cc. That's why I added them to the PR recently. Basically, it means someone touching the proto would need to build for Windows or Linux to generate the pb.h/cc, then switch to Android.

**Comment by Keenuts** [comment]
> I will spin it back to JSON (or YAML). I can keep most of the code, since the export mechanics is just a part of the PR.

If you do, JSON please. This won't be hand-written, just processed, so I'd be in favor of JSON.
The value I see in yaml is the less-verbose syntax for human writers. But JSON can be inlined and compacted a bit, it's less ambiguous

**Comment by mdagois** [comment]
> If you do, JSON please. This won't be hand-written, just processed, so I'd be in favor of JSON.
> The value I see in yaml is the less-verbose syntax for human writers. But JSON can be inlined and compacted a bit, it's less ambiguous

Yeah, I thought I would use this: https://github.com/nlohmann/json.

**Comment by Keenuts** [comment]
> > If you do, JSON please. This won't be hand-written, just processed, so I'd be in favor of JSON.
> > The value I see in yaml is the less-verbose syntax for human writers. But JSON can be inlined and compacted a bit, it's less ambiguous
> 
> Yeah, I thought I would use this: https://github.com/nlohmann/json.

Yeah, it's header only, so easy to integrate

File: src/proto/metrics_report.proto (hunk 3 above)
Comment (mdagois) [reply src/proto/metrics_report.proto]: Obsolete, as I moved to using JSON.

File: src/ppx/application.cpp (hunk 2 above)
Comment (mdagois) [reply src/ppx/application.cpp]: Fixed.

File: src/CMakeLists.txt (hunk 1 above)
Comment (mdagois) [reply src/CMakeLists.txt]: Obsolete, as I moved to JSON.

File: src/ppx/application.cpp [hunk 4]
```diff
@@ -1219,6 +1238,38 @@ int Application::Run(int argc, char** argv)
             mAverageFrameTime = static_cast<float>(mTimer.MillisSinceStart() / mFrameCount);
         }
 
+        // Update default metrics in the current metrics run.
+        if (mSettings.useMetrics && mMetrics.pCurrentRun != nullptr) {
+            PPX_ASSERT_NULL_ARG(mMetrics.pCpuFrameTime);
```
Comment (apazylbe) [inline src/ppx/application.cpp:1243]: Should we move this to a helper function? Run() is pretty long

File: include/ppx/application.h [hunk 5]
```diff
@@ -384,6 +387,17 @@ class Application
 
     bool IsXrEnabled() const { return mSettings.xr.enable; }
 
+    // Start a new metric run and returns it.
```
[the remark this answers is not in the document]
Comment (mdagois) [reply include/ppx/application.h]: Fixed!

File: include/ppx/application.h [hunk 6]
```diff
@@ -384,6 +387,17 @@ class Application
...
+    // Only one run is ever active at the same time.
+    // Default metrics are automatically added to the run: framerate, cpu_frame_time and frame_count.
+    // Additional ones may be added by the caller.
+    // The run is automatically exported and saved to disk when the application shuts down.
+    metrics::Run* StartMetricsRun(const char* pName);
+
+    // Stop the currently active run.
```
[the remark this answers is not in the document]
Comment (mdagois) [reply include/ppx/application.h]: Fixed!

File: include/ppx/metrics.h [hunk 7]
```diff
@@ -55,6 +56,26 @@ struct MetricMetadata
...
 ////////////////////////////////////////////////////////////////////////////////
 
+// A report contains runs and metrics information.
+// It is meant to be saved to disk.
+class Report final
+{
+    friend class Manager;
+
+public:
+    static constexpr const char* FILE_EXTENSION = ".json";
```
[the remark this answers is not in the document]
Comment (mdagois) [reply include/ppx/metrics.h]: Fixed!

File: include/ppx/metrics.h [hunk 8]
```diff
@@ -200,6 +227,9 @@ class Manager final
 
     Run* AddRun(const char* pName);
 
+    // Export all the runs and metrics information into a report.
```
[the remark this answers is not in the document]
Comment (mdagois) [reply include/ppx/metrics.h]: Fixed!

File: src/ppx/application.cpp (hunk 4 above)
Comment (mdagois) [reply src/ppx/application.cpp]: I moved the metrics system update code into its own function.

File: src/ppx/metrics.cpp [hunk 9]
```diff
@@ -19,6 +19,38 @@ namespace metrics {
...
+{
+}
+
+bool Report::WriteToFile(const char* pFilepath) const
+{
+    PPX_ASSERT_NULL_ARG(pFilepath);
+    std::ofstream outputFile(pFilepath, std::ofstream::out);
+    outputFile << mContent.dump(4) << std::endl;
+    outputFile.close();
+    return true;
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/metrics.cpp]: I removed the return value.

File: include/ppx/application.h [hunk 10]
```diff
@@ -416,6 +430,9 @@ class Application
     void   DestroyPlatformWindow();
     bool   IsRunning() const;
 
+    // Updates the metrics.
```
[the remark this answers is not in the document]
Comment (mdagois) [reply include/ppx/application.h:463]: Sorry, this was always meant to be a private function. I moved to the right place.

File: src/ppx/application.cpp [hunk 11]
```diff
@@ -1391,6 +1413,94 @@ float2 Application::GetNormalizedDeviceCoordinates(int32_t x, int32_t y) const
...
+        metadata.unit                    = "";
+        metadata.interpretation          = metrics::MetricInterpretation::NONE;
+        mMetrics.pFrameCount             = mMetrics.pCurrentRun->AddMetric<metrics::MetricCounter>(metadata);
+    }
+
+    mMetrics.resetFramerateTracking = true;
+    return mMetrics.pCurrentRun;
+}
+
+void Application::StopMetricsRun()
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp:1486]: I use it locally to test the metrics.
It just stops the current run, so it should not save the report.
You could have several runs in a single BigWheels execution.
Using the benchmark as an example:
1. Configure some knobs
2. Start a run
3. Record for 10 secs
4. Stop the run
5. Repeat 1 to 4 several times with different run names
6. Exit the app (the report with all runs is exported)
Let me know if that's unclear.

File: src/ppx/application.cpp [hunk 12]
```diff
@@ -1391,6 +1413,94 @@ float2 Application::GetNormalizedDeviceCoordinates(int32_t x, int32_t y) const
...
 }
 
+metrics::Run* Application::StartMetricsRun(const char* pName)
+{
+    PPX_ASSERT_MSG(mSettings.useMetrics, "Application::Settings::useMetrics must be set to true before using the metrics capabilities");
+    if (!mSettings.useMetrics) {
+        return nullptr;
+    }
+
+    PPX_ASSERT_MSG(mMetrics.pCurrentRun == nullptr, "a run is already active; stop it before starting another one");
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp:1456]: No, I'd rather have this treated as a fundamental error. If a run is started successfully, it should be stopped.

File: include/ppx/application.h [hunk 13]
```diff
@@ -232,6 +233,8 @@ struct ApplicationSettings
     bool        headless              = false;
     bool        enableImGui           = false;
     bool        allowThirdPartyAssets = false;
+    // Set to true to opt-in for metrics
```
Comment (Keenuts) [inline include/ppx/application.h:236]: nit: "." at the end
also, clang format stopped aligning variables?

File: src/ppx/application.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp:1486]: Originally, the metrics were developed with automated tests as a use case. In this context, having automatic runs and a dump of the report at the end makes sense. It is also good to test if the thing actually works locally. But, that's definitely not a limitation.

My expectations is that people using the metrics will change the API to match their needs. I think @Slumpwuffle already has this kind of plans.

Basically, the way I see it is that we create a centralized point in application.h to avoid the situation where every single sample reimplements the same mechanics for metrics. But, instead of trying to anticipate every potential use cases, we let people tweak and add what they need along the way.

Does that seem reasonnable?

File: include/ppx/application.h (hunk 13 above)
Comment (mdagois) [reply include/ppx/application.h]: Fixed.
And yes, clang-format stops aligning after the comment.

File: src/ppx/application.cpp [hunk 14]
```diff
@@ -1424,6 +1446,94 @@ float2 Application::GetNormalizedDeviceCoordinates(int32_t x, int32_t y) const
     return ndc;
 }
 
+metrics::Run* Application::StartMetricsRun(const char* pName)
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp]: Both are used in application.h. I don't have a strong preference. Switched to std::string here.

File: src/ppx/application.cpp [hunk 15]
```diff
@@ -680,6 +683,22 @@ void Application::DispatchShutdown()
 {
     Shutdown();
 
+    // Save the metrics report to disk
+    if (mSettings.useMetrics) {
+        // Build a unique name for the report
+        std::stringstream                                        reportFilename;
+        const std::chrono::time_point<std::chrono::system_clock> now          = std::chrono::system_clock::now();
+        const time_t                                             current_time = std::chrono::system_clock::to_time_t(now);
+        reportFilename << "report_" << current_time << metrics::Report::kFileExtension;
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp:692]: I just wanted to delete bitcoin wallets. I guess I won't.
I added the check + an error log if that ever happens.

File: src/ppx/application.cpp [hunk 16]
```diff
@@ -680,6 +683,22 @@ void Application::DispatchShutdown()
...
+
+        // Export the report from the metrics manager
+        metrics::Report report;
+        mMetrics.manager.Export(reportFilename.str().c_str(), &report);
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp]: I think const char* and std::string are actually the ones for UTF8, while wstring is implementation-dependent (and different on Windows and Linux) as far as I remember.
Anyway, the metrics API only uses C strings in the public API, which is why Export uses that too.

File: src/ppx/metrics.cpp [hunk 17]
```diff
@@ -215,6 +308,22 @@ Run* Manager::AddRun(const char* pName)
     return pRun;
 }
 
+void Manager::Export(const char* pName, Report* pOutReport) const
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/metrics.cpp]: You missed it. No ninja commit was done in your back. :)
You can change the C string to std::string if they offend you. I don't have any strong aversion to std::string in that case.

Export is used in application.cpp (Application::DispatchShutdown).

As for the signature, I was returning a result before, which is why the output report was passed as parameters. I changed the API and the other exports as well.

File: src/ppx/application.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Keenuts) [reply src/ppx/application.cpp:1491]: Not sure I follow the whole thing.

> Basically, the way I see it is that we create a centralized point in application.h to avoid the situation where every single sample reimplements the same mechanics for metrics

Sure

> But, instead of trying to anticipate every potential use cases, we let people tweak and add what they need along the way.

This would still be supported since you have the pCurrentRun accessible no?

Maybe I'm missing something, the main point I had was around the logic to prevent concurrent runs. Seems like this could be done by the manager no?

[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp:1491]: OK, misunderstanding then. :)
I thought you were asking for a way to export anytime.

I think I see your point now. There would be only one function that both starts and stops the current run?
I must admit I prefer intentional APIs.

An application using a button could easily manage the state, and they'd need to handle the state anyway to show a proper button label. However, once more applications use the metrics, we can tweak the API to match what's actually convenient or done the most often.

File: src/ppx/application.cpp [hunk 18]
```diff
@@ -680,6 +683,27 @@ void Application::DispatchShutdown()
 {
     Shutdown();
 
+    // Save the metrics report to disk
+    if (mSettings.useMetrics) {
+        // Build a unique name for the report
```
Comment (apazylbe) [inline src/ppx/application.cpp:688]: Optional: Personally I think it would be more readable if this was wrapped in a helper SaveMetrics. But I will leave that up to you, perhaps we should also move it to the Metrics Manager?

File: src/ppx/application.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (apazylbe) [reply src/ppx/application.cpp:1491]: I actually like the idea of moving some of this metrics code in Application to Manager, since that's in theory what it's supposed to do, and metrics might be used only by a few applications, so we can keep Application cleaner, and if someone was interested in seeing how metrics work they can find that in a more centralized place

File: src/ppx/application.cpp (hunk 18 above)
Comment (mdagois) [reply src/ppx/application.cpp]: I added a function SaveMetricsReportToDisk.

File: include/ppx/application.h [hunk 19]
```diff
@@ -409,6 +412,17 @@ class Application
 
     bool IsXrEnabled() const { return mSettings.xr.enable; }
 
+    // Starts a new metric run and returns it.
+    // Only one run is ever active at the same time.
+    // Default metrics are automatically added to the run: framerate, cpu_frame_time and frame_count.
+    // Additional ones may be added by the caller.
```
[the remark this answers is not in the document]
Comment (mdagois) [reply include/ppx/application.h:418]: > I am curious what this is supposed to look like in terms of updating the metrics, would each application have its own method similar to UpdateMetrics?

Yes, exactly. Each application can add whatever metrics they want to a run. Each run has default common metrics (framerate, frame time and frame count for now), but any other metrics is for the application to define. For example, let's say a foveation application wants to track image quality, it would 1) start a run, 2) add a custom "quality" metric to that run, 3) compute the image quality each frame (or whatever frequency needed) and 4) record the computed quality in the metric.

> The Start/Stop functions are not called by the parent class, meaning it's up to each application to call those. Meaning they are in control of the run lifecycle.

Yes, the applications control the runs. it is not really the lifecycle per-se. Runs as not deleted when they are stopped. This start/stop cycle cannot be handled by the parent class. An application might want to create several discrete runs, with different parameters. Also, applications will know when they load assets or do heavy work that should not be part of the run.

> So I see this as a default set of metrics you have for free, but then each application needs to manually call the manager to manager.addMetric, and handles them itself.

Yes, that was the exact original purpose. The application gets to opt-in for metrics and can then start runs, one at a time, because having several runs does not make sense anyway.

The pseudo-code is actually misleading. It is more like the following. The application does not interact with the manager, only with the runs it starts. Metrics are added to those runs, as not all runs would necessarily need the same metrics.

```
class MyApp : Application {
    Gauge customMetric;
    
    Setup() {
       settings.enableMetrics = true;
    }
    
    OnClick() {
       if(run_active) {
          StopRun();
       } else {
          Run run = StartRun();
          customMetric = run.AddMetricGauge();
       }
    }
    
    MainLoop() {
       customMetric .logValue(x);
    }
```

Again, as people use the metrics through the base application class, we can tweak the API. @Slumpwuffle already has some ideas, and since he's the earliest user, I am all in to let him shape the API.

File: src/ppx/application.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp:1498]: No, that's a misunderstanding of the intent of the manager. Applications can already use the manager directly. That's exactly what FishTornado does. In that setting, whether an application creates tons of runs upfront with a bunch of metrics, start recording values, pause, record again, that's all up to the application. A recording flag would have no meaning here, because the application is in full control.

The integration in the Application class is meant to cover the basic use case of most applications: a single active run at a time with a bunch of predefined metrics (which are recorded automatically) + the ability to add custom metrics (which the application records) + the automatic dump to disk. I believe it will actually cover 99% of the cases, but I of course could be wrong.

If that's not clear, maybe we should discuss it live. :)

File: include/ppx/application.h [hunk 20]
```diff
@@ -232,6 +233,8 @@ struct ApplicationSettings
     bool        headless              = false;
     bool        enableImGui           = false;
     bool        allowThirdPartyAssets = false;
+    // Set to true to opt-in for metrics.
+    bool useMetrics = false;
```
[the remark this answers is not in the document]
Comment (mdagois) [reply include/ppx/application.h]: I renamed it.

**Review by apazylbe — APPROVED** [review]
lgtm, thanks! 
I still believe it would be nicer to have most of metrics related code not in application, and have applications go through a manager, but that's my personal preference.

File: src/ppx/application.cpp [hunk 21]
```diff
@@ -1424,6 +1458,99 @@ float2 Application::GetNormalizedDeviceCoordinates(int32_t x, int32_t y) const
...
+}
+
+bool Application::HasActiveRun() const
+{
+    return mSettings.enableMetrics && mMetrics.pCurrentRun != nullptr;
+}
+
+void Application::UpdateMetrics()
+{
+    if (!mSettings.enableMetrics || mMetrics.pCurrentRun == nullptr) {
```
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp:1519]: Ah nice catch, I will add this to the next PR.

**Comment by mdagois** [comment]
> lgtm, thanks! I still believe it would be nicer to have most of metrics related code not in application, and have applications go through a manager, but that's my personal preference.

Yes, that's a perfectly valid point of view. Once we have two use cases (@Slumpwuffle + benchmark), I believe we'll have a clearer view of what's actually useful and we can refactor it.

File: src/ppx/application.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (mdagois) [reply src/ppx/application.cpp]: Actually, sorry, I made the change as I wanted to rename HasActiveRun to HasActiveMetricsRun as well.

**Review by Keenuts — APPROVED** [review]
Fine to revisit once we have more users to this API
